### PR TITLE
Kubelet raw metrics API proposal

### PR DESCRIPTION
**Pull request by timstclair**
Forked from discussion on https://github.com/kubernetes/kubernetes/pull/15691
Addresses https://github.com/kubernetes/kubernetes/issues/12483

cc/ @vishh @jimmidyson @dchen1107 @bgrant0607

### CODE REVIEW

File: docs/proposals/kubelet-raw-metrics-api.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  - `/namespaces/`
+  - `/namespaces/{namespace}`
+
+Additionally, all endpoints (except root discovery endpoint) support the
+following optional query parameters:
+
+- `start` - start time to return metrics from; type json encoded `time.Time`
+- `end` - end time to return metrics to; type json encoded `time.Time`
+- `count` - maximum number of stats to return in each ContainerMetrics instance;
+  type int
```
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: As @lavalamp says, this is "poor-man's paging" -- Is there a better way to structure this parameter to support proper paging in the future (e.g. should this be max total stats, as opposed to max per-container stats?)

File: docs/proposals/kubelet-raw-metrics-api.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+Future metrics types, such as derived metrics, will add a prefix to relevant
```
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: If we're reusing this API at the cluster level (served by API server) do we still want raw metrics to be the "default" (a.k.a. unprefixed) type of metrics?

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 1 above)
Comment (vishh) [inline docs/proposals/kubelet-raw-metrics-api.md]: I wonder if we can get rid of this parameter. What we ideally need is a way to downsample the data returned, to reduce the load on the higher layers and the network.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 2 above)
Comment (vishh) [inline docs/proposals/kubelet-raw-metrics-api.md]: We do not have a use case for exposing raw metrics at the cluster level as of now. 
If you are referring to the existing [cluster level derived metrics API](https://github.com/kubernetes/kubernetes/pull/15862/files#diff-72f39b225c34ea2b6b0d856c9ebd6a33L66), I agree that adopting the API schema in this proposal makes sense.

**Comment by vishh** [comment]
Thanks for the proposal @timstclair! Much appreciated!

**Comment by timstclair** [comment]
FYI, added a few notes about implementation.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 1 above)
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: I'd prefer to have this as a `step` between results. We could calculate a reasonable step depending on size of window between `start` & `end`.

File: docs/proposals/kubelet-raw-metrics-api.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+}
+type Pod struct {
+  ObjectMeta, TypeMeta
+  Containers []ContainerMetrics
+}
+
+// ContainerMetrics is a k8s wrapper around cAdvisor metrics
+type ContainerMetrics struct {
+  ObjectMeta, TypeMeta
+  Info cadvisorv2.ContainerInfo
```
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: I'd prefer all type information to be in Kubernetes & provide a converter from cadvisor API to Kubernetes API, even if that is a 1-1 field mapping. More independent of implementation & can be versioned separately.

File: docs/proposals/kubelet-raw-metrics-api.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+### Endpoints
+
+All endpoints are GET endpoints, rooted at `/apis/metrics/v1alpha1/`
+
+- `/` - discovery endpoint; type resource list
+- `/nodes` - host metrics; type `[]metrics.Node`
+  - `/nodes/localhost` - When served by kubelet, the only node provided is
+    `localhost`; type metrics.Node
+- `/pods` - All pod metrics across all namespaces; type `[]metrics.Pod`
```
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: Do we want to have endpoints for services & replication controllers too that could wrap up those metrics for pods backing those? Not for alpha1 probably but in time.

Also label & field selectors would be really useful here to be consistent with the other resource APIs. Again probably not alpha1 but we need to keep these in mind for future IMO.

File: docs/proposals/kubelet-raw-metrics-api.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+## Design
+
+Kubelet will start providing a versioned API, following convensions of the API
+server. The metrics API will be its own [API group](api-group.md), and may
+eventually be reused at the cluster-level (outside the scope of this proposal).
+
+### Endpoints
+
+All endpoints are GET endpoints, rooted at `/apis/metrics/v1alpha1/`
```
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: What are your thoughts for derived metrics in future? Separate endpoint or parameterized call to this endpoint?

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 2 above)
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: If we can have consistency that would be great.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 1 above)
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: @jimmidyson what do you mean by a step? Time step? In that case, we would need to validate that the raw metrics step evenly divides it.

@vishh is step & downsample the same thing? In either case, would we aggregate samples, or just drop samples not requested?

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 2 above)
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: Does that mean we should update the cluster level derived metrics API endpoints too? They look clunky compared to these, especially if we're going to be adding these in a separate API subgroup now?

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 1 above)
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: Yes a time step. IMO metrics should be reported at regular intervals for ease of use & understanding. 

I would drop samples not requested.

Comment (vishh) [inline docs/proposals/kubelet-raw-metrics-api.md]: Yes.
An usual query is something like "give me stats every 30 seconds for the
last five minutes".
I can express the duration with `start` and `end`. Adding `step` or `rate`
will help us express the downsampling rate.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 2 above)
Comment (vishh) [inline docs/proposals/kubelet-raw-metrics-api.md]: Yeah. The existing PR for cluster level metrics is still open. cc @mvdan who was working on that PR.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 3 above)
Comment (vishh) [inline docs/proposals/kubelet-raw-metrics-api.md]: The cadvisor APIs are also versioned right. What do we get by adding one more layer of versioning here?

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 2 above)
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: Agree, we definitely want consistency here. I think the metrics api group should include all metrics, both at the kubelet level & cluster level, including raw & derived metrics, and eventually RC & services as well.

I don't think this means that every endpoint needs to be implemented both on the kubelet & cluster, but I do think they should be consistent and have common paths where relevant.

What is the status on the cluster-level derived metrics? I don't see those endpoints in the code yet?

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 5 above)
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: Addressed below, but I think it should be a separate endpoint, unless we can fit them neatly into the same metrics data type.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 4 above)
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: This proposal was specifically targeting the kubelet API, and I don't think those make sense at that level. However, the goal is for this API to scale to the cluster level, and I think those metrics could be cleanly added to the current design. It's a good thing to keep in mind though.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 3 above)
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: It's worth noting that the API is currently completely self contained. I don't have a strong opinion here, but I bet someone else does :)

I think the important pieces are that everything is documented together in the same place. If the cadvisor API has separate documentation, I think that's a pretty strong argument for cloning the API.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 4 above)
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: Good point about the label & field selectors. Other common query parameters are `pretty` & `resourceVersion`

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 5 above)
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: Agreed.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 1 above)
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: OK so can we change `count`? I prefer `step` or `interval` to `rate` as that normally refers to rate of change when talking about metrics.

IMO this should be optional & a reasonable step value should be calculated from the requested window size.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 2 above)
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: I think it makes sense to make this one proposal PR including the whole metrics API group. Otherwise it's going to be hard to keep proposals in sync. We are already diverging.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 3 above)
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: IMO cadvisor is an implementation detail. Great that the cadvisor API is versioned, but I think the metrics API needs to be fully self-contained & versioned independently. Also using the cadvisor API directly would leak cadvisor details through to the unversioned representations as per the Kubernetes API conventions.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 2 above)
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: Ok. I'll merge it into that proposal later today.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 1 above)
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: Ok, I went with step. However, I disagree that the default should be variable based on the window size. I think it will be better to have a default (predictable) default size. For instance, if I make a request and get back stats ever 10 seconds, and then increase my window size by 30 seconds, I'd expect to get 3 more stats back, not get the same number spaced 11 seconds apart.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 3 above)
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: I'm sold. We'll mirror the cadvisor v2 APIs in the versioned metrics API, and also add ObjectMeta & TypeMeta to each object.

**Comment by timstclair** [comment]
Ok, merged this proposal into the existing DerivedMetrics proposal. I updated that proposal to be consistent with everything proposed here. Comment away!

**Comment by jimmidyson** [comment]
:+1: Nice work @timstclair!

**Comment by timstclair** [comment]
@mwielgus - What were the heapster requirements we discussed yesterday? I remember:

- pod labels
- pod status

Was there anything else?

**Comment by mwielgus** [comment]
@timstclair 
Pod status and container CPU/mem request/limit.

cc: @piosz

File: docs/proposals/compute-resource-metrics-api.md [hunk 6]
```diff
@@ -61,13 +77,53 @@ kube-ui-v1-nd7in           0.07 cores  130 MB
...
+The metrics API will be its own [API group](api-group.md), and is shared by the
+kubelet and cluster API.  All endpoints are GET endpoints, rooted at
+`/apis/metrics/v1alpha1/`
+
+- `/` - discovery endpoint; type resource list
+- `/nodes` - host metrics; type `[]metrics.Node`
+  - `/nodes/localhost` - When served by kubelet, the only node provided is
+    `localhost`; type metrics.Node
+  - `/nodes/{node}` - metrics for a specific node
+- `/derivedNodes` - host metrics; type `[]metrics.DerivedNode`
```
Comment (dchen1107) [inline docs/proposals/compute-resource-metrics-api.md]: Can we document here the difference between /nodes and /derivedNodes? Same for /pods?

From the document, i cannot see any difference.

**Comment by dchen1107** [comment]
LGTM overall, except one comment.

File: docs/proposals/compute-resource-metrics-api.md [hunk 7]
```diff
@@ -61,13 +77,53 @@ kube-ui-v1-nd7in           0.07 cores  130 MB
...
+following optional query parameters:
+
+- `start` - start time to return metrics from; type json encoded `time.Time`
+- `end` - end time to return metrics to; type json encoded `time.Time`
+- `step` - the time step between each stats sample; type int (seconds), default
+  10s, must be a multiple of 10s
+
+As well as the common query parameters:
+
+- `pretty` - pretty print the response
```
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: Can you educate me with an example response with this query parameter @timstclair ?

File: docs/proposals/compute-resource-metrics-api.md [hunk 8]
```diff
@@ -35,15 +35,31 @@ Documentation for other releases can be found at
 
 ## Goals
 
-Provide resource usage metrics on pods and nodes on the API server to be used
-by the scheduler to improve job placement, utilization, etc. and by end users
-to understand the resource utilization of their jobs. Horizontal and vertical
-auto-scaling are also near-term uses.
+Provide resource usage metrics on pods and nodes through the API server to be
+used by the scheduler to improve job placement, utilization, etc. and by end
```
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: nit: replace `job` with `pod` to be more explicit.

File: docs/proposals/compute-resource-metrics-api.md [hunk 9]
```diff
@@ -35,15 +35,31 @@ Documentation for other releases can be found at
...
+users to understand the resource utilization of their jobs. Horizontal and
+vertical auto-scaling are also near-term uses. Additionally, a subset of the
```
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: IIUC, in the near-term, the API server will expose a subset of the APIs that the kubelet will expose. To start with, the API server and kubelet metrics APIs will be mutually exclusive.

File: docs/proposals/compute-resource-metrics-api.md [hunk 10]
```diff
@@ -35,15 +35,31 @@ Documentation for other releases can be found at
...
+metrics API should be served directly from the kubelet.
+
+### API Requirements
+
+- Provide machine level metrics, all pod metrics (in single request), specific
```
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: AFAIK, we do not have a plan to surface raw metrics via the API server. This proposal does not seem to state that explicitly. May be we want to propose the API group as a common entity and then split the proposal into api-server specific and kubelet specific sections?

File: docs/proposals/compute-resource-metrics-api.md [hunk 11]
```diff
@@ -80,16 +136,67 @@ namespace or service aggregation, metrics at those levels could also be
...
+type Pod struct {
+  TypeMeta
+  ObjectMeta              // Should include pod name
+  Containers []Container
+}
+type Container struct {
+  TypeMeta
+  ObjectMeta              // Should include container name
+  Spec ContainerSpec      // Mirrors cadvisorv2.ContainerSpec
+  Stats []ContainerStats  // Mirrors cadvisorv2.ContainerStats
```
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: nit: Should we include `type ContainerStats cadvisorv2.ContainerStats` for the sake of completeness?

File: docs/proposals/compute-resource-metrics-api.md [hunk 12]
```diff
@@ -61,13 +77,53 @@ kube-ui-v1-nd7in           0.07 cores  130 MB
...
+- Unsupported paths return status not found (404)
+  - `/namespaces/`
+  - `/namespaces/{namespace}`
+
+Additionally, all endpoints (except root discovery endpoint) support the
+following optional query parameters:
+
+- `start` - start time to return metrics from; type json encoded `time.Time`
+- `end` - end time to return metrics to; type json encoded `time.Time`
+- `step` - the time step between each stats sample; type int (seconds), default
```
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: Should we mention that step is calculated from the timestamp of the metric point closest to `start`?

Essentially `start` and `end` might not map directly to metrics timestamps.

**Comment by vishh** [comment]
> @timstclair 
Pod status and container CPU/mem request/limit.

Pod Status is a versioned API object. Exposing such an object through an alternate API doesn't sound like a good idea. If the kubelet were to mirror the api-server APIs, that might be ok (in read-only mode). 
A kubelet level API that exposes Resource limits of containers are useful beyond the realm of this metrics proposal as well - for example, one of the current proposals for exposing resource limits to pods is through a kubelet API. 
Given that, I'd prefer not including resource limits in the metrics APIs.

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 1 above)
Comment (timstclair) [inline docs/proposals/kubelet-raw-metrics-api.md]: Added `count` back as there is a heapster use case for it.

cc/ @mwielgus

File: docs/proposals/compute-resource-metrics-api.md (hunk 6 above)
Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: Done.

File: docs/proposals/compute-resource-metrics-api.md (hunk 7 above)
Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: I think most of our get endpoints have this parameter, so I included it here. It's basically whether to format the response json with whitespace to be more human readable. If you query the API in browser, I think it defaults to true - try adding `?pretty=false` to see it off.

File: docs/proposals/compute-resource-metrics-api.md (hunk 9 above)
Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: Thinking more about this, I'm worried that this will get confusing, especially if the API server and kubelet APIs diverge any more. I almost wonder if we want separate api groups for each (e.g. `kubelet-metrics`, or maybe just `kubelet` or `node`). On the one hand, we want consistency so keeping them together makes sense, on the other hand it should be clear which is served by which component. Separate groups also prevents reuse of common types across the APIs. Another option would be to prefix kubelet-only types with "kubelet" (e.g. s/pods/kubeletPods/), but this could get messy if we decide to start serving them at the cluster level.

Thoughts?

File: docs/proposals/compute-resource-metrics-api.md (hunk 10 above)
Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: See above comment.

File: docs/proposals/compute-resource-metrics-api.md (hunk 11 above)
Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: I was trying to avoid copying the entire cadvisor API here. Is it useful to add one more level of depth here? For now I'll just add a link to the cadvisor API.

File: docs/proposals/compute-resource-metrics-api.md (hunk 9 above)
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: In the future, I can see kubelet might exposing `derived` metrics, which will help us scale heapster as well. I don't see why we will have two types if the versions match in that case.

Perhaps, we can state the general APIs in the beginning, and then explicitly state which component will be exposing these APIs?

File: docs/proposals/compute-resource-metrics-api.md (hunk 11 above)
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: SGTM

File: docs/proposals/compute-resource-metrics-api.md [hunk 13]
```diff
@@ -61,13 +77,57 @@ kube-ui-v1-nd7in           0.07 cores  130 MB
...
+  10s, must be a multiple of 10s
+- `count` - maximum number of stats to return in each ContainerMetrics instance;
```
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: Why is this necessary? @piosz @mwielgus

File: docs/proposals/compute-resource-metrics-api.md (hunk 12 above)
Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: Added note under `start`

File: docs/proposals/compute-resource-metrics-api.md (hunk 8 above)
Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: done

File: docs/proposals/compute-resource-metrics-api.md (hunk 9 above)
Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: Do you mean in the documentation?

Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: This comment has now been suppressed by the recent rebase :(

Yeah. I mean better documentation could satisfy my concerns. But in general, if we have a need for diverging the data format between the kubelet and api-server, we might have to abandon this idea. Since this is versioned and we are starting in an `alpha` phase, I'd say lets roll with the current proposal, with better documentation. WDYT?

Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: Ok, works for me.

For future reference, if I make changes in separate commits (don't squash) does it prevent github from collapsing comment threads?

File: docs/proposals/compute-resource-metrics-api.md (hunk 10 above)
Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: This goes back to an early question of whether we want the "raw" metrics to be the default (unprefixed) type. If they're not being exposed at the cluster level, I think it makes this more of an issue. I think we should add back the "raw" prefix.

File: docs/proposals/compute-resource-metrics-api.md (hunk 9 above)
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: Yeah. Multiple commits followed by a squash after LGTM is the recommended
approach for PRs that have lots of comments.

File: docs/proposals/compute-resource-metrics-api.md (hunk 10 above)
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: +1

Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: Done.

File: docs/proposals/compute-resource-metrics-api.md [hunk 14]
```diff
@@ -61,13 +77,60 @@ kube-ui-v1-nd7in           0.07 cores  130 MB
...
+only be available through the kubelet. The types of metrics are detailed
+[below](#schema). All endpoints are GET endpoints, rooted at
+`/apis/metrics/v1alpha1/`
+
+- `/` - discovery endpoint; type resource list
+- `/rawNodes` - raw host metrics; type `[]metrics.RawNode`
+  - `/rawNodes/localhost` - The only node provided is `localhost`; type
+    metrics.Node
+- `/derivedNodes` - host metrics; type `[]metrics.DerivedNode`
+  - `/nodes/{node}` - derived metrics for a specific node
```
Comment (jimmidyson) [inline docs/proposals/compute-resource-metrics-api.md]: `derivedNodes`

File: docs/proposals/kubelet-raw-metrics-api.md (hunk 1 above)
Comment (jimmidyson) [inline docs/proposals/kubelet-raw-metrics-api.md]: What is the use case? What is the behaviour if available metrics are greater than requested `count`? Trim from beginning or end of window?

**Comment by dchen1107** [comment]
LGTM

**Comment by timstclair** [comment]
@k8s-bot test this

**Comment by timstclair** [comment]
squashed

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

File: docs/proposals/compute-resource-metrics-api.md [hunk 15]
```diff
@@ -80,16 +143,73 @@ namespace or service aggregation, metrics at those levels could also be
...
+// Raw metrics are only available through the kubelet API.
+type RawNode struct {
+  TypeMeta
+  ObjectMeta              // Should include node name
+  Machine ContainerMetrics
+  SystemContainers []ContainerMetrics
+}
+type RawPod struct {
+  TypeMeta
+  ObjectMeta              // Should include pod name
```
Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: @timstclair: I just remembered that network usage is at the pod level and volumes disk usage is also at the pod level. So in addition to container metrics, we will have pod level metrics too.

Comment (timstclair) [inline docs/proposals/compute-resource-metrics-api.md]: Would we move those stats out of [ContainerStats](https://github.com/timstclair/cadvisor/blob/master/info/v2/container.go#L105) and into RawPod, add additional NetworkStats into the RawPod, or add a new Network stats type?

Comment (vishh) [inline docs/proposals/compute-resource-metrics-api.md]: Yes. Its either that or we can pre-define an infrastructure container and
associate all pod level stats to that container. Even in that case, volume
fs usage cannot belong to any containers.
